### PR TITLE
Add operation level interceptor support

### PR DESCRIPTION
**Pull request by AmaliMatharaarachchi**
### Issues

Fixes https://github.com/wso2/product-microgateway/issues/2606

### Automation tests
 - Unit tests added: No
 - Integration tests added: Yes

### Tested environments

Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: router/src/main/resources/interceptor/lib/interceptor.lua [hunk 1]
```diff
@@ -366,16 +379,28 @@ end
...
     local meta = response_handle:streamInfo():dynamicMetadata():get(LUA_FILTER_NAME)
     local shared_info = meta and meta[SHARED_INFO_META_KEY]
     if not shared_info then
         -- no shared info found, request interceptor flow is not executed (eg: enforcer validation failed)
         -- TODO: (renuka) check again, if we want to go response flow if enforcer validation failed, have to set request info metadata from enforcer
         return
     end
+    local method = shared_info[REQUEST.REQ_HEADERS][":method"]
+    if method == nil then
+        response_handle:logErr("Header :method is missing in shared info's request headers.")
```
Comment (renuka-fernando) [inline router/src/main/resources/interceptor/lib/interceptor.lua:394]:
```suggestion
        response_handle:logErr("Error getting header ':method' in response flow interceptor. The header is not found in shared info's request headers.")
```

File: adapter/internal/interceptor/interceptor.go [hunk 2]
```diff
@@ -69,11 +76,10 @@ var (
...
-			responseHeaders={{.RespFlowInclude.ResponseHeaders}}, responseBody={{.RespFlowInclude.ResponseBody}}, responseTrailers={{.RespFlowInclude.ResponseTrailers}}}
-{{else}}local resp_flow = {}{{end}} {{/* if resp_flow disabled no need req info in resp path */}}
-{{if or .ReqFlowInclude.InvocationContext .RespFlowInclude.InvocationContext}}
+{{if .ResponseFlowEnable}} {{/* resp_flow details are required in req flow if request info needed in resp flow */}}
+local resp_flow_list = {  {{ range $key, $value := .ResponseFlow }} {{ $key }}={invocationContext={{$value.Include.InvocationContext}}, requestHeaders={{$value.Include.RequestHeaders}}, requestBody={{$value.Include.RequestBody}}, requestTrailer={{$value.Include.RequestTrailer}},
+		responseHeaders={{$value.Include.ResponseHeaders}}, responseBody={{$value.Include.ResponseBody}}, responseTrailers={{$value.Include.ResponseTrailers}}}, {{ end }}}
+{{else}}local resp_flow_list = {}{{end}} {{/* if resp_flow disabled no need req info in resp path */}}
```
[the remark this answers is not in the document]
Comment (AmaliMatharaarachchi) [reply adapter/internal/interceptor/interceptor.go]: Thanks for the suggestion. fixed https://github.com/wso2/product-microgateway/pull/2602/commits/f8131414492b760fc2d74eeade0f7c2b220e778e

File: adapter/internal/interceptor/interceptor.go [hunk 3]
```diff
@@ -19,22 +19,29 @@ package interceptor
...
 
 //Interceptor hold values used for interceptor
 type Interceptor struct {
-	Context              *InvocationContext
-	RequestExternalCall  *HTTPCallConfig
-	ResponseExternalCall *HTTPCallConfig
-	ReqFlowInclude       *RequestInclusions
-	RespFlowInclude      *RequestInclusions
+	Context            *InvocationContext
+	RequestFlowEnable  bool
```
Comment (praminda) [inline adapter/internal/interceptor/interceptor.go:30]: How about `EnableRequestFlow` or `IsRequestFlowEnabled`

Comment (AmaliMatharaarachchi) [reply adapter/internal/interceptor/interceptor.go]: fixed with https://github.com/wso2/product-microgateway/pull/2602/commits/f3dc0441e1de5f7cc7e9e0636ab90dcc6a62b1ea

**Review by renuka-fernando — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[failed] dev-deployment-v2 : 20220125.15

**Comment by choreo-cicd** [comment]
[failed] dev-deployment-v2 : 20220125.16

**Comment by choreo-cicd** [comment]
[succeeded] stage-deployment-v2 : 20220126.1

**Comment by choreo-cicd** [comment]
[succeeded] prod-deployment-v2 : 20220128.1

**Pull request merged**